### PR TITLE
Fixes issue where FAIL_OVERFLOW results in tight loop of retrying.

### LINKED ISSUE #30258: [BUG] BufferedProducer drops in send rate after some time
**Describe the bug**

While running the stress suite we noticed that the sending rate drops dramatically after time using azure-messaing-eventhubs 5.13.0-beta.1 https://search.maven.org/artifact/com.azure/azure-messaging-eventhubs/5.13.0-beta.1/jar

image

***Exception or Stack Trace***
N/A

**To Reproduce**
* Create an application publishing 500 events every second.

**Expected behavior**
The rate remains constant.

**Setup (please complete the following information):**
 - IDE: [e.g. IntelliJ]
 - Library/Libraries: [e.g. com.azure:azure-core:1.16.0 (groupId:artifactId:version)]
 - Java version: [e.g. 8] 
 - App Server/Environment: [e.g. Tomcat, WildFly, Azure Function, Apache Spark, Databricks, IDE plugin or anything special]
 - Frameworks: [e.g. Spring Boot, Micronaut, Quarkus, etc] 

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Bug Description Added
- [x] Repro Steps Added
- [x] Setup information Added

### PR DESCRIPTION
**Pull request by conniey**
# Description

- Fixes issue where FAIL_OVERFLOW results in tight loop of retrying.
- Uses retryWhen to apply the AMQP retry policy when trying to publish events downstream.

Fixes #30258

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedPartitionProducer.java [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+                return;
+            }
+
+            if (isClosed.get()) {
+                sink.error(new IllegalStateException(String.format("Partition publisher id[%s] was "
+                    + "closed between flushing events and now. Cannot enqueue events.", partitionId)));
+                return;
+            }
+
+            final Sinks.EmitResult emitResult1 = eventSink.tryEmitNext(eventData);
```
Comment (liukun-msft) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedPartitionProducer.java:121]: nit: emitResult?

**Review by liukun-msft — APPROVED** [review]
LGMT! Since we merge buffer this into main branch, do we need to add docs and samples for BufferedProducer?

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedPartitionProducer.java (hunk 1 above)
Comment (conniey) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedPartitionProducer.java:121]: Good catch.

**Pull request merged**